src/sixel.5
.TH SIXEL 5 "May 21, 2014"
.SH NAME
sixel \- SIXEL format
.SH DESCRIPTION
SIXEL is a bitmap graphics format for terminals and printers introduced
by Digital Equipment Corp. (DEC).
Its data scheme is represented as a terminal-friendly escape sequence.
So if you want to show a SIXEL image file, all you have to do is "cat"
it to your terminal.
.br

.TP 5
Terminal requirements:
Several of DEC VT series, VT240/VT241/VT330/VT340/VT282/VT284/VT286/VT382
.br
RLogin (Japanese terminal emulator)
.br
tanasinn (Works with firefox)
.br
mlterm (Works on each of X, win32/cygwin and framebuffer version)
.br
XTerm (configured with --enable-sixel-graphics and lanuched with "-ti 340" option)
.br
yaft / yaftx (Works on framebuffer / X11 environment)
.br
DECterm
.br
Kermit for DOS
.br
WRQ Reflection
.br
ZSTEM

.SH "SEE ALSO"
.BR "img2sixel"(1), " sixel2png"(1), "
.LP
All About SIXELs, Sep 29, 1990:
.IP
.br
ftp://ftp.cs.utk.edu/pub/shuford/terminal/all_about_sixels.txt
.LP
Displaying Sixel Image Files, 2014:
.IP
.br
http://rullf2.xs4all.nl/sg/doc.html
.LP

.\" end of man page



NEWS
                  ------------------------------
                   What's new in libsixel-1.7 ?
                  ------------------------------

* core: Upgrade stb_image to 2.12.

* core: Upgrade stb_image_write.h to v1.02.

* core: introduce canonical,fast decoder to prevent segmentation faults

* python: add row level API functions to python interface

* img2sixel: fix broken -S option

* img2sixel: fix broken animated GIF resizing feature

* example: add example/drawing

* example: add example/python

* converters: Makefile tweak from yacp(thanks to @fd00)
  https://github.com/fd00/yacp/blob/896e9bd819b373e8e99fefcec12bd85c73e7f123/libsixel/libsixel-1.6.0-1bl1.src.patch


                  ------------------------------
                   What's new in libsixel-1.6 ?
                  ------------------------------
* img2sixel: Add -R(--gri-limit) option.

* configure: Fix check for gdImageCreateFromPngPtr availability(thanks to @hattya).

* core: Upgrade stb_image to 2.10.

* core: Upgrade stb_image_write.h to v1.01.

* libsixel-config: Add --datarootdir option to libsixel-config.

* img2sixel: Solve encoder option confliction between -m/-e/-I/-b.

* core: Introduce new API for output object, sixel_output_set_gri_arg_limit().

* core: Add new API: sixel_encoder_encode_bytes().


                  ------------------------------
                   What's new in libsixel-1.5 ?
                  ------------------------------
* img2sixel: Add -k(--insecure) option.

* img2sixel: Add -B(--bgcolor) option.

* core: Introduce new API for dither object, sixel_dither_set_transparent().

* core: Respect transparency of paletted images(GIF/PNG).

* core: Introduce API set for generic image maniplation:
    sixel_frame_get_pixels()
    sixel_frame_ref()
    sixel_frame_unref()
    sixel_frame_init()
    sixel_frame_get_pixels()
    sixel_frame_get_palette()
    sixel_frame_get_width()
    sixel_frame_get_height()
    sixel_frame_get_ncolors()
    sixel_frame_get_pixelformat()
    sixel_frame_get_transparent()
    sixel_frame_get_multiframe()
    sixel_frame_get_delay()
    sixel_frame_get_frame_no()
    sixel_frame_get_loop_no()
    sixel_frame_strip_alpha()
    sixel_frame_resize()
    sixel_frame_clip()

* core: Introduce API set for load/write well-known image formats:
    sixel_helper_load_image_file()
    sixel_helper_write_image_file()

* core: Add more high level APIs (File-to-File conversion API):
    sixel_encoder_create()
    sixel_encoder_ref()
    sixel_encoder_unref()
    sixel_encoder_set_cancel_flag()
    sixel_encoder_setopt()
    sixel_encoder_encode()
    sixel_decoder_create()
    sixel_decoder_ref()
    sixel_decoder_unref()
    sixel_decoder_setopt()
    sixel_decoder_decode()

* core: Introduce new APIs for semantic error handling:
    sixel_helper_set_additional_message()
    sixel_helper_get_additional_message()
    sixel_helper_format_error()

* core: Introduce custom allocator APIs:
    sixel_allocator_new()
    sixel_allocator_ref()
    sixel_allocator_unref()
    sixel_allocator_malloc()
    sixel_allocator_calloc()
    sixel_allocator_realloc()
    sixel_allocator_free()

* core: Add new constructor methods that returns SIXELSTATUS for semantic error handling:
    sixel_output_new()
    sixel_dither_new()
    sixel_frame_new()
    sixel_encoder_new()
    sixel_decoder_new()
    sixel_allocator_new()
    sixel_output_new()
    sixel_dither_new()
    sixel_frame_new()
    sixel_encoder_new()
    sixel_decoder_new()

* core: Improve processing quality of loading paletted PNG images.

* core: Improve processing quality and performance of loading GIF images.

* core: Update stb_image.h to version 2.06

* core: Update stb_image_write.h to version 0.94

* Add Python interface

* Add Perl interface (alpha)

* Add PHP interface (alpha)

* Add Ruby interface (alpha)

* Add new build helper program: libsixel-config

* A lot of bug fixes and minor improvements.
  Thanks to @msmhrt, @ttdoda, and @msmhrt.


                  ------------------------------
                   What's new in libsixel-1.4 ?
                  ------------------------------
* core: SIXEL encoder supports various pixel formats(thanks to @uobikiemukot).

* core: Supports high-color(15bpp output) mode(thanks to @arakiken).

* core: More high compressed output mode is introduced(thanks to @arakiken).

* core: Fix typoed API(thanks to @waywardmonkeys).

* core: Fix a minor performance bug(thanks to @arakiken).

* img2sixel: New option -D(--pipe-mode) is introduced.
             With this option, img2sixel continuously read source images from
             stdin.

* img2sixel: Introduce -I(--high-color) option, high-color(15bpp) output mode
             is added(thanks to @arakiken).

* img2sixel: New option -E(--encode-policy) is added. high-compressed
             mode(-E size) is introduced(thanks to @arakiken).

* img2sixel: New option -b(--builtin-palette) is added. built-in support of
             VT-340 palette is added(thanks to @vrtsds).

* img2sixel: Performance inprovement for loading 8bit indexed PNG.

* tests: 'make test' achieves 90% coverage.

* A lot of bug fixes and minor improvements.
  Thanks to @obache, @ttdoda, @vrtsds, and @msmhrt.


                   ------------------------------
                   What's new in libsixel-1.3 ?
                  ------------------------------

* Now libcurl is automatically linked to img2sixel.
  If you don't want to build with libcurl, you should configure with
  --without-libcurl option.

* Add libjpeg integration(linked automatically to img2sixel).

* Add libpng integration(linked automatically to img2sixel/sixel2png).

* Introduce -t(--palette-type) option.
  img2sixel can output HLS paletted sixel with "-t hls".
  Add new API sixel_output_set_palette_type.

* Introduce -v(--verbose) option.

* Add new API sixel_dither_set_body_only.

* Now img2sixel can load SIXEL as input data.

* Some bug fixes and minor improvements.

php/sixel.c
/*
 * Copyright (c) 2015 Hayaki Saito
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* $ Id: $ */ 

#include "php_sixel.h"

#if HAVE_SIXEL

/* {{{ Class definitions */

/* {{{ Class SixelEncoder */

static zend_class_entry * SixelEncoder_ce_ptr = NULL;

/* {{{ Methods */


/* {{{ proto object __construct()
   */
PHP_METHOD(SixelEncoder, __construct)
{
	zend_class_entry * _this_ce;
	zval * _this_zval;



	if (ZEND_NUM_ARGS()>0)  {
		WRONG_PARAM_COUNT;
	}


	do {
		SIXELSTATUS status;
				sixel_encoder_t *encoder;
				zval *value;
				status = sixel_encoder_new(&encoder, NULL);
				if (SIXEL_FAILED(status)) {
#if 0
					zend_throw_exception_ex(zend_exception_get_default(), 1,
											"sixel_encoder_new() failed: %s (%s:%d)",
											sixel_helper_format_error(status),
											__FILE__, __LINE__);
#endif
				} else {
					value = emalloc(sizeof(zval));
					ZVAL_RESOURCE(value, (long)encoder);
					zend_update_property(_this_ce, getThis(),
										 "encoder", sizeof("encoder") - 1, value);
				}
	} while (0);
}
/* }}} __construct */



/* {{{ proto object __destruct()
   */
PHP_METHOD(SixelEncoder, __destruct)
{
	zend_class_entry * _this_ce;

	zval * _this_zval = NULL;



	if (zend_parse_method_parameters(ZEND_NUM_ARGS() TSRMLS_CC, getThis(), "O", &_this_zval, SixelEncoder_ce_ptr) == FAILURE) {
		return;
	}

	_this_ce = Z_OBJCE_P(_this_zval);


	do {
		zval *encoder;
				encoder = zend_read_property(_this_ce, getThis(),
											 "encoder", sizeof("encoder") - 1, 1);
				sixel_encoder_unref((sixel_encoder_t *)Z_RESVAL_P(encoder));
				efree(encoder);
	} while (0);
}
/* }}} __destruct */



/* {{{ proto void setopt(string opt[, string arg])
   */
PHP_METHOD(SixelEncoder, setopt)
{
	zend_class_entry * _this_ce;

	zval * _this_zval = NULL;
	const char * opt = NULL;
	int opt_len = 0;
	const char * arg = NULL;
	int arg_len = 0;



	if (zend_parse_method_parameters(ZEND_NUM_ARGS() TSRMLS_CC, getThis(), "Os|s", &_this_zval, SixelEncoder_ce_ptr, &opt, &opt_len, &arg, &arg_len) == FAILURE) {
		return;
	}

	_this_ce = Z_OBJCE_P(_this_zval);


	do {
		SIXELSTATUS status;
				zval *encoder;
				encoder = zend_read_property(_this_ce, getThis(),
											 "encoder", sizeof("encoder") - 1, 1);
				status = sixel_encoder_setopt((sixel_encoder_t *)Z_RESVAL_P(encoder), *opt, arg);
#if 0
				if (SIXEL_FAILED(status) {
					zend_throw_exception_ex(zend_exception_get_default(), 1,
											"sixel_encoder_encode() failed: %s (%s:%d)",
											sixel_helper_format_error(status),
											__FILE__, __LINE__);
				}
#endif
	} while (0);
}
/* }}} setopt */



/* {{{ proto void encode(string filename)
   */
PHP_METHOD(SixelEncoder, encode)
{
	zend_class_entry * _this_ce;

	zval * _this_zval = NULL;
	const char * filename = NULL;
	int filename_len = 0;



	if (zend_parse_method_parameters(ZEND_NUM_ARGS() TSRMLS_CC, getThis(), "Os", &_this_zval, SixelEncoder_ce_ptr, &filename, &filename_len) == FAILURE) {
		return;
	}

	_this_ce = Z_OBJCE_P(_this_zval);


	do {
		SIXELSTATUS status;
				zval *encoder;
				encoder = zend_read_property(_this_ce, getThis(),
											 "encoder", sizeof("encoder") - 1, 1);
				status = sixel_encoder_encode((sixel_encoder_t *)Z_RESVAL_P(encoder), filename);
#if 0
				if (SIXEL_FAILED(status) {
					zend_throw_exception_ex(zend_exception_get_default(), 1,
											"sixel_encoder_encode() failed: %s (%s:%d)",
											sixel_helper_format_error(status),
											__FILE__, __LINE__);
				}
#endif
	} while (0);
}
/* }}} encode */


static zend_function_entry SixelEncoder_methods[] = {
	PHP_ME(SixelEncoder, __construct, NULL, /**/ZEND_ACC_PUBLIC | ZEND_ACC_CTOR)
	PHP_ME(SixelEncoder, __destruct, NULL, /**/ZEND_ACC_PUBLIC)
	PHP_ME(SixelEncoder, setopt, SixelEncoder__setopt_args, /**/ZEND_ACC_PUBLIC)
	PHP_ME(SixelEncoder, encode, SixelEncoder__encode_args, /**/ZEND_ACC_PUBLIC)
	{ NULL, NULL, NULL }
};

/* }}} Methods */

static void class_init_SixelEncoder(void)
{
	zend_class_entry ce;

	INIT_CLASS_ENTRY(ce, "SixelEncoder", SixelEncoder_methods);
	SixelEncoder_ce_ptr = zend_register_internal_class(&ce);
}

/* }}} Class SixelEncoder */

/* }}} Class definitions*/

/* {{{ sixel_functions[] */
zend_function_entry sixel_functions[] = {
	{ NULL, NULL, NULL }
};
/* }}} */


/* {{{ sixel_module_entry
 */
zend_module_entry sixel_module_entry = {
	STANDARD_MODULE_HEADER,
	"sixel",
	sixel_functions,
	PHP_MINIT(sixel),     /* Replace with NULL if there is nothing to do at php startup   */ 
	PHP_MSHUTDOWN(sixel), /* Replace with NULL if there is nothing to do at php shutdown  */
	PHP_RINIT(sixel),     /* Replace with NULL if there is nothing to do at request start */
	PHP_RSHUTDOWN(sixel), /* Replace with NULL if there is nothing to do at request end   */
	PHP_MINFO(sixel),
	PHP_SIXEL_VERSION, 
	STANDARD_MODULE_PROPERTIES
};
/* }}} */

#ifdef COMPILE_DL_SIXEL
ZEND_GET_MODULE(sixel)
#endif


/* {{{ PHP_MINIT_FUNCTION */
PHP_MINIT_FUNCTION(sixel)
{
	class_init_SixelEncoder();

	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_MSHUTDOWN_FUNCTION */
PHP_MSHUTDOWN_FUNCTION(sixel)
{

	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_RINIT_FUNCTION */
PHP_RINIT_FUNCTION(sixel)
{
	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_RSHUTDOWN_FUNCTION */
PHP_RSHUTDOWN_FUNCTION(sixel)
{
	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_MINFO_FUNCTION */
PHP_MINFO_FUNCTION(sixel)
{
	php_printf("PHP interface to libsixel\n");
	php_info_print_table_start();
	php_info_print_table_row(2, "Version",PHP_SIXEL_VERSION " (alpha)");
	php_info_print_table_row(2, "Released", "2015-06-23");
	php_info_print_table_row(2, "CVS Revision", "$Id: $");
	php_info_print_table_row(2, "Authors", "Hayaki Saito 'saitoha@me.com' (developer)\n");
	php_info_print_table_end();
	/* add your stuff here */

}
/* }}} */

#endif /* HAVE_SIXEL */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
